File: basket/backup_server.py
"""
Lightweight static file server with a backup save endpoint.
Run from project root and open http://localhost:8000/
It serves the project files and accepts POST /save-backup with JSON body to save backups/backups.json
"""
import http.server
import socketserver
import os
import json
from urllib.parse import urlparse

PORT = 8000
BACKUP_DIR = 'backups'
BACKUP_FILE = os.path.join(BACKUP_DIR, 'backup.json')

class Handler(http.server.SimpleHTTPRequestHandler):
    def do_POST(self):
        parsed = urlparse(self.path)
        if parsed.path == '/save-backup':
            length = int(self.headers.get('Content-Length', 0))
            body = self.rfile.read(length)
            try:
                data = json.loads(body.decode('utf-8'))
            except Exception as e:
                self.send_response(400)
                self.end_headers()
                self.wfile.write(b'Invalid JSON')
                return
            try:
                os.makedirs(BACKUP_DIR, exist_ok=True)
                with open(BACKUP_FILE, 'w', encoding='utf-8') as f:
                    json.dump(data, f, ensure_ascii=False, indent=2)
                self.send_response(200)
                self.end_headers()
                self.wfile.write(b'OK')
                print(f"Saved backup to {BACKUP_FILE}")
            except Exception as e:
                self.send_response(500)
                self.end_headers()
                self.wfile.write(b'Write failed')
                print('Failed to write backup:', e)
        else:
            self.send_response(404)
            self.end_headers()
            self.wfile.write(b'Not found')

if __name__ == '__main__':
    print(f"Serving on port {PORT}. Use Ctrl-C to stop.")
    print(f"Backups will be written to {os.path.abspath(BACKUP_FILE)}")
    with socketserver.TCPServer(("", PORT), Handler) as httpd:
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print('\nStopping server')
            httpd.server_close()

File: basket/backup_server.js
/*
Node.js backup server for the Basket project.

- Serves static files from project root (index.html, src/, assets/, ...)
- Provides POST /save-backup to receive JSON and save it to backups/backup.json
- Serves saved backup at /backups/backup.json for client-side auto-restore

Usage:
  1) Install Node.js (https://nodejs.org/) if not present
  2) In project folder run: npm install
  3) Start server: npm start

This is equivalent to the previous Python server but implemented in JavaScript (Node + Express).
*/

const express = require('express');
const fs = require('fs');
const path = require('path');
const bodyParser = require('body-parser');

const app = express();
const PORT = process.env.PORT || 8000;
const BACKUP_DIR = path.join(__dirname, 'backups');
const BACKUP_FILE = path.join(BACKUP_DIR, 'backup.json');

// Serve static project files from current directory
app.use(express.static(__dirname));

// parse JSON bodies
app.use(bodyParser.json({limit: '10mb'}));

app.post('/save-backup', (req, res) => {
  const data = req.body;
  if(!data || typeof data !== 'object'){
    return res.status(400).send('Invalid JSON');
  }
  try{
    if(!fs.existsSync(BACKUP_DIR)) fs.mkdirSync(BACKUP_DIR, { recursive: true });
    fs.writeFileSync(BACKUP_FILE, JSON.stringify(data, null, 2), { encoding: 'utf8' });
    console.log(`Saved backup to ${BACKUP_FILE}`);
    res.status(200).send('OK');
  }catch(err){
    console.error('Failed to write backup:', err);
    res.status(500).send('Write failed');
  }
});

// Optional health endpoint
app.get('/health', (req, res) => res.send({ok:true}));

app.listen(PORT, () => console.log(`Backup server running at http://localhost:${PORT}/`));
